### ISSUE TITLE
[ENH] Budget Optimizer implementation proposal

### ISSUE THREAD

**Issue by felipeangelimvieira** [felipeangelimvieira/prophetverse#199]
# Context

One of the most requested features in MMM is a budget optimizer. It has been on Prophetverse’s roadmap for over a year, and now that the library is more mature and its interface more robust, adding this feature would complete the core functionality for MMM. This issue proposes an implementation.

Budget optimization can be viewed as optimization of the exogenous features passed to the model after training. For example, Meridian uses a grid-search strategy similar to hyperparameter tuning. Because the hill-shaped response curves and other standard marketing-investment effects are smooth and well-behaved, non-linear optimization (e.g., via SciPy) can be effective—an approach adopted by PyMC-Marketing.

The objective here is to propose **a unified interface for budget optimization** that can support a wide variety of strategies.

# Proposal

## Key components of a budget optimizer

A budget optimizer typically involves these elements:

- **MMM model**: the fitted model which, given inputs, produces forecasts at the original data frequency.  
- **Utility/Objective function**: a scalar function to be maximized; it takes the model’s forecasts  and inputs and computes a single metric (e.g., ROI, revenue uplift).  
- **Constraints**: rules on spending, which can take many forms, such as:  
  - a total budget to distribute across a subset of channels  
  - a fixed monthly budget for a specific channel  
  - a maximum spend for a channel in a given month  
  - a minimun target response
  Handling different constraint types is tricky because each optimization backend may require its own interface and only supports a limited feature set.  
- **Decision variables/Parametrization**: the exogenous inputs (columns and dates in the `X` DataFrame) that the optimizer will adjust.  The decision variables can change: it can be possible to optimize the entire matrix of dates x channels, or keep the shape of daily investment per channel and optimize channel share.
- **Optimization backend**: an external solver (SciPy, a JAX-based optimizer, etc.) that takes a function and its parameters and returns an optimal solution.

The initial idea was to support multiple backends: scipy, optax, optuna, hyperactive. However, the SLSQP method available in `scipy.optimize` is unique in the sense of allowing a versatile set on constraints (on inputs and outputs), bounds and objective functions. In this sense, this first version will focus entirely on Scipy backend.

We can organize these into three families of objects:

### 1. Constraints

**What they are**  
Encapsulations of any ​rule​ you want to impose on spending or response – e.g. total budget ceilings, channel-level caps, or minimum KPI thresholds. Each constraint returns exactly what SciPy’s `minimize(..., constraints=…)` expects: a dict with  
- `type`: `"eq"` or `"ineq"`  
- `fun`: the function $g(x)$ whose zero (for `"eq"`) or non-negativity (for `"ineq"`) defines the feasible region  
- `jac`: its Jacobian for efficient gradient-based solving  

**Key classes**  
- `BaseConstraint`  
- `SharedBudgetConstraint` (equality constraint ensuring $\sum_{\text{channels, dates}}x = \text{total budget}$) 
- `MinimumTargetResponse` (inequality constraint enforcing model output ≥ target response)  

Beyond these, users can subclass `BaseConstraint` to implement things like per-channel monthly caps, ROI floors, or any custom linear/non-linear rule.

---

### 2. Objective function

**What it is**  
A single scalar measure to ​maximize​ or ​minimize​: ROI, total revenue uplift, or simply total spend. Under the hood it calls your trained Prophetverse model’s predictive engine, aggregates over the forecast horizon, and returns a real‐valued score.

**Key classes**  
- `BaseOptimizationObjective`  
- `MaximizeROI` (maximizes $\frac{\sum_{\tau\in\text{horizon}}\mathbb{E}[y_\tau]}{\sum x}$, negated for minimization)  
- `MaximizeKPI` (maximizes total expected KPI over horizon)  
- `MinimizeBudget` (finds minimal spend reaching other constraints)  

You can extend `BaseOptimizationObjective` for any custom utility—e.g. weighted multi‐objective scores or risk‐adjusted returns.

---

### 3. Parametrization transform

**What it is**  
A mapping between the ​flat​ decision vector SciPy sees and richer, structured “channels × time” matrices. This lets you reparameterize the problem (for smoother search or fewer variables) while enforcing things like fixed temporal shares.

**Key classes**  
- `BaseParametrizationTransformation`  
- `IdentityTransform` (no change; decision variable = raw spend time series)  
- `InvestmentPerChannelTransform` (optimizes only per‐channel sums, then redistributes by each day’s historical share)  

Transforms implement  
```python
.fit(X, horizon, columns)              # capture shape, shares
.transform(x0)                         # map full guess → low-dim param
.inverse_transform(low_dim_param)      # reconstruct full spend time series
```

### 4. Budget Optimizer

**What it is**  
The “glue” that stitches model, objective, constraints, and transform into a single call.

**Init arguments**  
- `objective: BaseOptimizationObjective`  
  The objective function to be optimized (e.g. `MaximizeROI`, `MaximizeKPI`, `MinimizeBudget`).  
- `constraints: List[BaseConstraint]`  
  A list of constraint objects (e.g. `SharedBudgetConstraint`, `MinimumTargetResponse`).  
- `parametrization_transform: Optional[BaseParametrizationTransformation]`  
  How to re-parametrize the decision vector (default `IdentityTransform`).  
- `method: str`  
  SciPy optimization method (default `"SLSQP"`).  
- `tol: float`  
  Tolerance for termination (passed to SciPy).  
- `bounds: Union[List[tuple], Dict[str, tuple]]`  
  Bounds on decision variables—either a list of `(low, high)` pairs or a dict mapping each channel to its bound.  
- `options: dict`  
  Extra options for the optimizer (e.g. max iterations).  
- `callback: callable`  
  Optional function called at each iteration with the current `x`.

**Optimize arguments**  
- `model`  
  A fitted Prophetverse model instance.  
- `X: pd.DataFrame`  
  Input data containing historical and (to-be-optimized) future spend.  
- `horizon: pd.Index`  
  The forecast dates (rows of `X`) over which to optimize spend.  
- `columns: pd.Index`  
  Which exogenous columns (channels) in `X` to treat as decision variables.

---

1. **Prepare**  
   - Extract initial spend ($x_0$) from `X.loc[horizon, columns]`  
   - Fit & apply your `parametrization_transform`  
   - Build SciPy constraints list from each `BaseConstraint(...)` → `dict(type, fun, jac)`  
   - Wrap the objective under the inverse transform for SciPy  

2. **Solve**  
   - Call `scipy.optimize.minimize(fun, x0, jac=…, bounds=…, constraints=…, method="SLSQP", …)`  

3. **Unpack**  
   - Inverse-transform the solution back to the full daily × channel matrix  
   - Return a copy of the original `X` with optimized rows in `horizon` replaced by the new spend  

**Key classes**  
- `BaseBudgetOptimizer`  
- `BudgetOptimizer` (SciPy-based SLSQP implementation)

**Comment by felipeangelimvieira** [comment]
@marrov FYI
